ntoskrnl/inbv/inbv.c

```diff
@@ -793,13 +792,29 @@ InbvRotationThread(
     }
     X = ProgressBarLeft + 2;
     Y = ProgressBarTop + 2;
+    InbvReleaseLock();
 
-    while (InbvDisplayState == INBV_DISPLAY_STATE_OWNED && PltRotBarStatus != 3)
+    while (InbvDisplayState == INBV_DISPLAY_STATE_OWNED)
     {
-        RotBarType = RotBarSelection;
-        InbvReleaseLock();
+        InbvAcquireLock();
+        // if (PltRotBarStatus == 1), do palette fading-in
+        // if (PltRotBarStatus == 2), do rotation bar animation
+
+        if (PltRotBarStatus == 3)
+        {
+            /* Stop the thread */
+            InbvReleaseLock();
+            break;
+        }
+
+        if (PltRotBarStatus > 3)
+        {
+            /* Unknown command, wait for another one */
+            InbvReleaseLock();
+            goto Wait;
+        }
 
-        if (RotBarType == RB_SQUARE_CELLS)
+        if (RotBarSelection == RB_SQUARE_CELLS)
         {
             Delay.QuadPart = -800000; // 0.08 sec
```

Comment:
Why do you move the Delay and Total variables initializations inside the while() loop now (in your previous patch version they were outside, done when initializing Index).

Reply:
This was done for better security of thread (for example, if `RotBarSelection` was changed from `RB_PROGRESS_BAR` to `RB_SQUARE_CELLS` on the fly, the code wouldn't draw something weird in this situation).